sdks/python/apache_beam/io/gcp/bigquery.py

```diff
@@ -1641,3 +1644,314 @@ def process(self, unused_element, signal):
+  """Read data from BigQuery.
+
+    PTransform:ReadAllFromBigQueryRequest->Rows
+
+    This PTransform uses a BigQuery export job to take a snapshot of the table
+    on GCS, and then reads from each produced JSON file.
+
+    It is recommended not to use this PTransform for streaming jobs on
+    GlobalWindow, since it will not be able to cleanup snapshots.
+
+  Args:
+    gcs_location (str, ValueProvider): The name of the Google Cloud Storage
+      bucket where the extracted table should be written as a string or
+      a :class:`~apache_beam.options.value_provider.ValueProvider`. If
+      :data:`None`, then the temp_location parameter is used.
+    project (str): The ID of the project containing this table.
+    validate (bool): If :data:`True`, various checks will be done when source
+      gets initialized (e.g., is table present?).
+    coder (~apache_beam.coders.coders.Coder): The coder for the table
+      rows. If :data:`None`, then the default coder is
+      _JsonToDictCoder, which will interpret every row as a JSON
+      serialized dictionary.
+    schema: The schema to be used if the BigQuery table to write has to be
+      created. This can be either specified as a 'bigquery.TableSchema' object
+      or a single string  of the form 'field1:type1,field2:type2,field3:type3'
+      that defines a comma separated list of fields. Here 'type' should
+      specify the BigQuery type of the field. Single string based schemas do
+      not support nested fields, repeated fields, or specifying a BigQuery
+      mode for fields (mode will always be set to 'NULLABLE').
+    kms_key (str): Experimental. Optional Cloud KMS key name for use when
+      creating new temporary tables.
+   """
+  def __init__(
+      self,
+      gcs_location=None,
+      project=None,
+      validate=False,
+      coder=None,
+      schema=None,
+      flatten_results=True,
+      kms_key=None):
+    if gcs_location:
+      if not isinstance(gcs_location, (str, unicode, ValueProvider)):
+        raise TypeError(
+            '%s: gcs_location must be of type string'
+            ' or ValueProvider; got %r instead' %
+            (self.__class__.__name__, type(gcs_location)))
+
+      if isinstance(gcs_location, (str, unicode)):
+        gcs_location = StaticValueProvider(str, gcs_location)
+
+    if schema is None:
+      raise ValueError("Should provide schema.")
+
+    self.coder = coder
+    self.gcs_location = gcs_location
+    self.project = project
+    self.validate = validate
+    self.flatten_results = flatten_results
+    self.coder = coder or _JsonToDictCoder
+    self.schema = bigquery_tools.get_dict_table_schema(schema)
+    self.kms_key = kms_key
+
+  def _get_destination_uri(self, temp_location):
+    """Returns the fully qualified Google Cloud Storage URI pattern where the
+    extracted table should be written to.
+    """
+    file_pattern = '{}/bigquery-table-dump-*.json'
+
+    if self.gcs_location is not None:
```

Comment:
Can you elaborate on this one please?

Reply:
You're doing `gcs_location.get` because gcs location is a value provider - which we use for templates.

Templates allow you to create a template file which represents an expanded pipeline, and submit it to dataflow along with runtime parameters, that get passed to RuntimeValueProviders.

RuntimeValueProviders are filled in only at pipeline runtime(i.e. in `process`, `start_bundle`, `finish_bundle`), not at pipeline construction (i.e. in `expand`).

This code will work for parameters passed at construction time, but it will not support templates/runtimevalueproviders. If you want to support RuntimeValueProviders, you should only call `get` on them in runtime functions (`process`, `finish_bundle`, `start_bundle`, others).